packages/cli/src/lib/setup/setupUpgrade.ts

```diff
@@ -552,101 +594,111 @@ function Upgrade(options) {
-                console.log(`Update ${adapter} from @${ioInstalled.common.version} to @${version}`);
+                console.log(`Update ${adapter} from @${ioInstalled.common!.version} to @${version}`);
                 // Get the adapter from web site
-                const { packetName, stoppedList } = await install.downloadPacket(sources, `${adapter}@${version}`);
+                // @ts-expect-error it could also call processExit internally but we want change it in future anyway
+                const { packetName, stoppedList } = await this.install.downloadPacket(sources, `${adapter}@${version}`);
                 await finishUpgrade(packetName);
-                await install.enableInstances(stoppedList, true);
+                await this.install.enableInstances(stoppedList, true);
             } else {
                 return console.error(`Unable to get version for "${adapter}".`);
             }
         }
-    };
+    }
 
     /**
      * Upgrade the js-controller
      *
-     * @param {string} repoUrl
-     * @param {boolean} forceDowngrade
-     * @param {boolean} controllerRunning
-     * @return {Promise<void>}
+     * @param repoUrl
+     * @param forceDowngrade
+     * @param controllerRunning
      */
-    this.upgradeController = async function (repoUrl, forceDowngrade, controllerRunning) {
-        if (!repoUrl || typeof repoUrl !== 'object') {
+    async upgradeController(
+        repoUrlOrObject: string,
+        forceDowngrade: boolean,
+        controllerRunning: boolean
+    ): Promise<void> {
+        let sources: Record<string, any>;
+        if (!repoUrlOrObject || !tools.isObject(repoUrlOrObject)) {
             try {
-                const result = await getRepository(repoUrl, params);
+                const result = await this.getRepository(repoUrlOrObject, this.params);
                 if (!result) {
-                    return console.warn(`Cannot get repository under "${repoUrl}"`);
+                    return console.warn(`Cannot get repository under "${repoUrlOrObject}"`);
                 }
-                repoUrl = result;
+                sources = result;
             } catch (err) {
-                return processExit(err);
+                return this.processExit(err);
             }
+        } else {
+            sources = repoUrlOrObject;
         }
 
         const installed = fs.readJSONSync(`${tools.getControllerDir()}/io-package.json`);
         if (!installed || !installed.common || !installed.common.version) {
             return console.error(
-                `Host "${hostname}"${hostname.length < 15 ? ''.padStart(15 - hostname.length) : ''} is not installed.`
+                `Host "${this.hostname}"${
+                    this.hostname.length < 15 ? ''.padStart(15 - this.hostname.length) : ''
+                } is not installed.`
             );
         }
-        if (!repoUrl[installed.common.name]) {
+        if (!sources[installed.common.name]) {
```

Comment:
I'd also put this into a constant, both `installed.common.name`  and `sources[installed.common.name]`